src/ReflectionFile.php

```diff
@@ -50,6 +50,9 @@ class ReflectionFile
      */
     public function __construct($fileName, $topLevelNodes = null)
     {
+        if (!is_string($fileName)) {
```

Comment:
1. huge PRs are pain to review (this PR is becoming larger and larger)
2. the solution you've chosen for Scrutinizer CI reported problem isn't correct one

The Scrutinizer CI told that multiple types could come in and you've restricted to type check instead of actually investigating places from where multiple types could come in and deciding on fixing these places instead. This could be a lots of changes and that's why I've suggested another PR.

P.S.
You don't need to put ALL fixing in single PR. Better to send smaller manageable PRs. Right now this PR has large discussion that already is hard to track what's going on.